## Table of Content

1. **[Pandas](#pandas)**

2. **[Data Structures](#structures)**
    
3. **[Pandas Series](#series)**
    - 3.1 - [Creating a Series](#creatingS)
    - 3.2 - [Manipulating Series](#manipulatingS)

4. **[Pandas DataFrames](#dataframes)**
    - 4.1 - [Creating DataFrames](#creatingDF)
    - 4.2 - [Manipulating DataFrames](#manipulatingDF)

5. **[Reading Data from Different Sources](#reading_data)**


<a id="pandas"> </a>
# 1. Pandas

Pandas contain data structures and data manipulation tools designed for data cleaning and analysis.
<br><br>
                        While pandas adopt much code from NumPy, the difference is that Pandas is designed for tabular, heterogeneous data. NumPy, by difference, is best suited for working with homogeneous numerical array data.<br><br>
                         The name Pandas is derived from the term 'panel data' (an econometrics term for multidimensional structured data sets).
                

**How to install pandas?**<br>
1. You can use-<br>
`!pip install pandas`<br>
2. You can import it as 'pd'<br>
import pandas as pd

 Import the pandas library; the following convention is used

In [1]:
import pandas as pd

So from now on, we will use `pd` instead of pandas. 

<a id="structures"> </a>
# 2. Data Structures

Pandas has two data structures as follows:<br>
1. A Series is a 1-dimensional labelled array that can hold data of any type (such as integer, string, boolean, float, python objects). Its axis labels are collectively called an index.<br>
2. A DataFrame is a 2-dimensional labelled data structure with columns. It supports multiple data types.

<a id="series"> </a>
# 3. Pandas Series

Pandas Series is a one-dimensional labelled array capable of holding any data type. However, a series is a sequence of similar data types, similar to an array, list, or column in a table. <br><Br>It will assign a labelled index to each item in the pd.Series. By default, each item will receive an index label from 0 to N, where N is the length of the Series minus one.

<a id="creatingS"> </a>
## 3.1 Creating a Series

**1. To create a numeric series** 

In [2]:
# create a numeric series
numbers = range(1,100,5)

# covert it to series
pd.Series(numbers)

0      1
1      6
2     11
3     16
4     21
5     26
6     31
7     36
8     41
9     46
10    51
11    56
12    61
13    66
14    71
15    76
16    81
17    86
18    91
19    96
dtype: int64

The output also gives the data type of the series as `int64`

Furthermore,  note that by default, each item will receive an index label from 0 to N, where N is the length of the Series minus one.

<b>In python, the row names are called index.</b>

**2. To create an object series** 

In [3]:
# create an object series
string = "Hi" , "How" ,"are", "you", "?"


# create a series from the above list
pd.Series(string)

0     Hi
1    How
2    are
3    you
4      ?
dtype: object

The output gives the data type of the series as `object`

**3. To create a series by giving both numeric and string values** 

In [4]:
# create a Series with an arbitrary list
pd.Series([345, 'London', 34.5, -34.45, 'Happy Birthday'])

0               345
1            London
2              34.5
3            -34.45
4    Happy Birthday
dtype: object

Here the numeric values are treated as object.

**4. To set index values for a series**

In [5]:
# declare a list of marks
marks = [60, 89, 74, 86]

# declare a list of subjects
subject = ["Maths", "Science", "English" , "Social Science"]

# create a series from the above list of marks with subjects as its index names
# index: adds the index 
marks_series = pd.Series(marks, index = subject) 
marks_series

Maths             60
Science           89
English           74
Social Science    86
dtype: int64

The index is added using the argument `index=`. The data type of the series continues to be numeric.

**5. To print the values and index of the Series**

In [6]:
# print the index of the series
marks_series.index

Index(['Maths', 'Science', 'English', 'Social Science'], dtype='object')

In [7]:
# prints the values of the series
marks_series.values

array([60, 89, 74, 86])

**6. To create a series from a dictionary**

In [8]:
# declare a dictionary
data = {'Maths': 60, 'Science': 89, 'English': 76, 'Social Science': 86}

# create a series from the dictionary
# the dictionary keys are the index names
# the dictionary values are the series values
pd.Series(data)

Maths             60
Science           89
English           76
Social Science    86
dtype: int64

On passing a `dict`, the index in the resulting Series will have the dict’s keys in given order.

**7. To create a series using library `numpy`**

In [9]:
# import the numpy library
import numpy as np

# declare the variable 'sequence' using linspace()
sequence = np.linspace(0,10, 5)

# print the array 'sequence'
sequence

array([ 0. ,  2.5,  5. ,  7.5, 10. ])

We see that we obtain an array. We can now convert it to a Series.

In [10]:
# create a series from the above sequence
pd.Series(sequence)

0     0.0
1     2.5
2     5.0
3     7.5
4    10.0
dtype: float64

**8. A series with missing values**

If we pass a key that is not defined, then its value will be `NaN`.

In [11]:
# create a list of subjects
subjects = ["Maths", "Science", "Art and Craft" , "Social Science"]

# declare a dictionary
data = {'Maths': 60, 'Science': 89, 'English': 76, 'Social Science': 86}

# create a series from the dictionary
marks_series = pd.Series(data, index = subjects)

# print series
print(marks_series)

Maths             60.0
Science           89.0
Art and Craft      NaN
Social Science    86.0
dtype: float64


<a id="manipulatingS"> </a>
### 3.2 Manipulating Series 

**1. To check for null values using `.isnull`**

In [12]:
# check for nulls in the data
marks_series.isnull()

Maths             False
Science           False
Art and Craft      True
Social Science    False
dtype: bool

`False` indicates that the value is not null.

**2. To check for null values using `.notnull`**

In [13]:
# check for nulls in the data
marks_series.notnull()

Maths              True
Science            True
Art and Craft     False
Social Science     True
dtype: bool

` True` indicates that the value is not null.

**3. To know the subjects in which marks score is more than 75**

In [14]:
# check for marks more than 75
marks_series[marks_series > 75]

Science           89.0
Social Science    86.0
dtype: float64

**4. To assign 68 marks to 'Art and Craft'**

In [15]:
# assign 68 marks to 'Art and Craft'
marks_series["Art and Craft"] = 68

In [16]:
# print the series
marks_series

Maths             60.0
Science           89.0
Art and Craft     68.0
Social Science    86.0
dtype: float64

**5. To check whether Maths marks are 73**

In [17]:
# check whether Maths marks are 73
marks_series.Maths == 73

False

In [18]:
# or you may use
marks_series["Maths"] == 73

False

**6. To create a series by generating numpy random numbers**

In [19]:
# create the numbers series 
# generate a sequence of 15 random numbers using random()
# round(): rounds off the number to nearest integer
num = pd.Series(np.random.random(15)*10).round()

**7. To find the square of the numbers series**

In [20]:
# declare a variable square 
# variable 'square' contains the 
square = pd.Series(num*num)
square.index = [num]
square

3.0       9.0
9.0      81.0
7.0      49.0
4.0      16.0
3.0       9.0
3.0       9.0
4.0      16.0
8.0      64.0
9.0      81.0
10.0    100.0
6.0      36.0
5.0      25.0
3.0       9.0
2.0       4.0
4.0      16.0
dtype: float64

**8. To assign index name and object name**

In [21]:
# assign object name
square.name = 'Square'

# assign index name
square.index.name = 'Number'

# print the series
square

3.0       9.0
9.0      81.0
7.0      49.0
4.0      16.0
3.0       9.0
3.0       9.0
4.0      16.0
8.0      64.0
9.0      81.0
10.0    100.0
6.0      36.0
5.0      25.0
3.0       9.0
2.0       4.0
4.0      16.0
Name: Square, dtype: float64

**Note:** A Series’s index can be altered in-place by assignment

From the output, it is not clear that the index column is labeled, to check whether it is labeled let us print it

We use `Series.index` to print the index.

In [22]:
# print the index
square.index

MultiIndex([( 3.0,),
            ( 9.0,),
            ( 7.0,),
            ( 4.0,),
            ( 3.0,),
            ( 3.0,),
            ( 4.0,),
            ( 8.0,),
            ( 9.0,),
            (10.0,),
            ( 6.0,),
            ( 5.0,),
            ( 3.0,),
            ( 2.0,),
            ( 4.0,)],
           name='Number')

From `name='Number'`, it is seen that the column is labeled.

**9. Add a number 5 to every element of the series**

In [23]:
# add 5 to each element
square + 5

3.0      14.0
9.0      86.0
7.0      54.0
4.0      21.0
3.0      14.0
3.0      14.0
4.0      21.0
8.0      69.0
9.0      86.0
10.0    105.0
6.0      41.0
5.0      30.0
3.0      14.0
2.0       9.0
4.0      21.0
Name: Square, dtype: float64

The number 5 has been added to all the variables.

**10. To extract a value specifying the index**

In [24]:
# obtain the value having index 7
square[7]

7.0    49.0
Name: Square, dtype: float64

All values with index as '7' are obtained

**11. To extract a range of values specifying the location**

In [25]:
# extract values having the location 
# obtain the values from the 3rd position till the 6th position
# indexing starts from 0
# the 7th position is not included
square[3:7]

4.0    16.0
3.0     9.0
3.0     9.0
4.0    16.0
Name: Square, dtype: float64

**12. Usage of `.iloc`**

We use `.iloc` to get the values of the specified index of numbers 

In [26]:
# obtain the value in the 5th position
square.iloc[5]

9.0

In [27]:
# obtain the values from the 3rd till the 8th position
# indexing starts from 0
# the 7th position is not included
square.iloc[3:9]

4.0    16.0
3.0     9.0
3.0     9.0
4.0    16.0
8.0    64.0
9.0    81.0
Name: Square, dtype: float64

**13. Sorting a numeric series**

In [28]:
# create a pandas series
age = pd.Series([23, 45, np.nan, 41, 23, 34, 55, np.nan, 34, 20])

# print values
age

0    23.0
1    45.0
2     NaN
3    41.0
4    23.0
5    34.0
6    55.0
7     NaN
8    34.0
9    20.0
dtype: float64

In [29]:
# ascending order
# sort_values: sorts the values
# ascending : if specified as True, it sorts values in ascending order (default value is True)
age.sort_values(ascending = True)

9    20.0
0    23.0
4    23.0
5    34.0
8    34.0
3    41.0
1    45.0
6    55.0
2     NaN
7     NaN
dtype: float64

In [30]:
# arrange in descending order
# sort_values: sorts the values
# ascending : if specified as True, it sorts values in ascending order (default value is True)
# set ascending to False to sort the values in ascending order
age.sort_values(ascending = False)

6    55.0
1    45.0
3    41.0
8    34.0
5    34.0
4    23.0
0    23.0
9    20.0
2     NaN
7     NaN
dtype: float64

**14. Sorting a categorical series**

In [31]:
# create a pandas series
string_values = pd.Series(["a", "j", "d", "f", "t", "a"])

# print the values
string_values

0    a
1    j
2    d
3    f
4    t
5    a
dtype: object

In [32]:
# ascending order
# sort_values: sorts the values
# ascending : if specified as True, it sorts values in ascending order (default value is True)
string_values.sort_values(ascending = True)

0    a
5    a
2    d
3    f
1    j
4    t
dtype: object

In [33]:
# descending order
# sort_values: sorts the values
# ascending : if specified as True, it sorts values in ascending order (default value is True)
# set ascending to False to sort the values in ascending order
string_values.sort_values(ascending = False)

4    t
1    j
3    f
2    d
5    a
0    a
dtype: object

**15. Sorting based on index**

In [34]:
# recall the series 'square'
square

3.0       9.0
9.0      81.0
7.0      49.0
4.0      16.0
3.0       9.0
3.0       9.0
4.0      16.0
8.0      64.0
9.0      81.0
10.0    100.0
6.0      36.0
5.0      25.0
3.0       9.0
2.0       4.0
4.0      16.0
Name: Square, dtype: float64

In [35]:
# sort in ascending order based on index
# sort_index: sorts the series based on the index
# ascending : if specified as True, it sorts values in ascending order (default value is True)
square.sort_index(ascending = True)

2.0       4.0
3.0       9.0
3.0       9.0
3.0       9.0
3.0       9.0
4.0      16.0
4.0      16.0
4.0      16.0
5.0      25.0
6.0      36.0
7.0      49.0
8.0      64.0
9.0      81.0
9.0      81.0
10.0    100.0
Name: Square, dtype: float64

In [36]:
# sort in descending order based on index
# sort_index: sorts the series based on the index
# ascending : if specified as True, it sorts values in ascending order (default value is True)
# set ascending to False to sort the values in ascending order
square.sort_index(ascending = False)

10.0    100.0
9.0      81.0
9.0      81.0
8.0      64.0
7.0      49.0
6.0      36.0
5.0      25.0
4.0      16.0
4.0      16.0
4.0      16.0
3.0       9.0
3.0       9.0
3.0       9.0
3.0       9.0
2.0       4.0
Name: Square, dtype: float64

**16. Rank a Series**

In [37]:
# recall the marks_series
marks_series

Maths             60.0
Science           89.0
Art and Craft     68.0
Social Science    86.0
dtype: float64

In [38]:
# rank the marks in ascending order
# rank(): ranks the values of a series 
marks_series.rank()

Maths             1.0
Science           4.0
Art and Craft     2.0
Social Science    3.0
dtype: float64

In [39]:
# rank the marks in ascending order
# rank(): ranks the values of a series 
# ascending : if specified as True, it sorts values in ascending order (default value is True)
# set ascending to False to sort the values in ascending order
marks_series.rank(ascending = False)

Maths             4.0
Science           1.0
Art and Craft     3.0
Social Science    2.0
dtype: float64

<a id="dataframes"> </a>
## 4. Pandas DataFrames

 A DataFrame is a tabular representation of data containing an ordered collection of columns, each of which can be a different type (such as numeric, string, boolean). <br><br>
                        The DataFrame has both row and column index; it can be thought of as a dict of Series all sharing the same index. In a data frame, the data is stored as one or more two-dimensional blocks rather than a list, dict, or some other collection of one-dimensional arrays. <br><br>
                        While a DataFrame is physically two-dimensional, it can be used to represent higher dimensional data in a tabular format using hierarchical indexing.
                   

<a id="creatingDF"> </a>
### 4.1 Creating DataFrames

**1. Creating a data frame a dictionary**

In [40]:
# create a dictionary
data = {'Subject': ['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art'],
        'Marks': (45, 65, 78, 65, 80, 78),
        'GPA': [2.5, 3.0, 3.5, 2.0, 4.0, 4.0]}

# create the dataframe using DataFrame()
df_marks = pd.DataFrame(data)

# print the dataframe
print(df_marks)

    Subject  Marks  GPA
0     Maths     45  2.5
1   History     65  3.0
2   Science     78  3.5
3   English     65  2.0
4  Georaphy     80  4.0
5       Art     78  4.0


**Note:** Like Series, the resulting DataFrame is assigned index automatically. And the 'Marks' values are in a tuple. 

**2. Another way to create dataframe from dictionary**

In [41]:
# creare the dictionary
data = [{'Subject': 'Maths', 'Marks': 45, 'GPA':2.5},
        {'Subject':'History', 'Marks':65, 'GPA':3.0},
        {'Subject':'Science', 'Marks':78, 'GPA':3.5},
        {'Subject':'English', 'Marks':65, 'GPA':2.0},
        {'Subject':'Georaphy', 'Marks':80, 'GPA':4.0},
        {'Subject':'Art', 'Marks':78, 'GPA':4.0}]

# create the dataframe using DataFrame()
df_marks = pd.DataFrame(data)

# print the dataframe
print(df_marks)

    Subject  Marks  GPA
0     Maths     45  2.5
1   History     65  3.0
2   Science     78  3.5
3   English     65  2.0
4  Georaphy     80  4.0
5       Art     78  4.0


**3. To create dataframe from lists**

In [42]:
# declare the list 'Subject'
Subject = ['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art']

# declare the list 'Marks'
Marks = [45, 65, 78, 65, 80, 78]

# declare the list 'CPA'
GPA = [2.5, 3.0, 3.5, 2.0, 4.0, 4.0]

In [43]:
# create a DataFrame from the lists
# index: specifies the index names
df_marks = pd.DataFrame([Subject, Marks, GPA], index = ['Subject','Marks','GPA'])

# print the DataFrame
df_marks

0        1        2        3         4    5
Subject  Maths  History  Science  English  Georaphy  Art
Marks       45       65       78       65        80   78
GPA        2.5        3      3.5        2         4    4

However to want a vertical dataframe so we use `.T`. The 'T' stands for transpose.

In [44]:
# transpose the DataFrame
df_marks.T

Subject Marks  GPA
0     Maths    45  2.5
1   History    65    3
2   Science    78  3.5
3   English    65    2
4  Georaphy    80    4
5       Art    78    4

**4. To create dataframe from series**

In [45]:
# declare the series 'Subject'
Subject = pd.Series(['Maths', 'History', 'Science', 'English', 'Georaphy', 'Art'])

# declare the series 'Marks'
Marks = pd.Series([45, 65, 78, 65, 80, 78])

# declare the series 'GPA'
GPA = pd.Series([2.5, 3.0, 3.5, 2.0, 4.0, 4.0])

In [46]:
# create a DataFrame from the Series
# index: specifies the index names
df_marks = pd.DataFrame([Subject, Marks, GPA], index = ['Subject','Marks','GPA'])

# print the DataFrame
df_marks

0        1        2        3         4    5
Subject  Maths  History  Science  English  Georaphy  Art
Marks       45       65       78       65        80   78
GPA        2.5        3      3.5        2         4    4

However to want a vertical dataframe so we use `.T`. The 'T' stands for transpose.

In [47]:
# transpose the DataFrame
df_marks.T

Subject Marks  GPA
0     Maths    45  2.5
1   History    65    3
2   Science    78  3.5
3   English    65    2
4  Georaphy    80    4
5       Art    78    4

**Remark:** Assign a name to the data frame and then use `.T` to transpose it.

**5. To read data from csv file**

In [48]:
df_basic_info = pd.read_csv("basic_info.csv")

In [49]:
type(df_basic_info)

pandas.core.frame.DataFrame

On checking the data type, we notice it is read as pandas data frame.

In [50]:
print(df_basic_info)

    Age  Weight (in kg)  Height (in m)
0    45              60           1.35
1    12              43           1.21
2    54              78           1.50
3    26              65           1.21
4    68              50           1.32
5    21              43           1.52
6    10              32           1.65
7    57              34           1.61
8    75              23           1.24
9    32              21           1.52
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85


**6. To print head of the data**

In [51]:
# display the first 5 rows of the DataFrame
# head(): displays the first 5 rows
# to display more rows, for example 15, use head(15)
# the default value is 5
df_basic_info.head()

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50
3   26              65           1.21
4   68              50           1.32

By default, the `.head()` will display **first** five rows. However, we can set the desired number of rows to be displayed.

Say we want to see the first 9 rows, we write the number 9 in the parentheses.

In [52]:
# display 9 rows
df_basic_info.head(9)

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50
3   26              65           1.21
4   68              50           1.32
5   21              43           1.52
6   10              32           1.65
7   57              34           1.61
8   75              23           1.24

**7. To print tail of the data**

In [53]:
# to display the last 5 rows
df_basic_info.tail()

Age  Weight (in kg)  Height (in m)
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

By default, the `.tail()` will display **last** five rows. However, we can set the desired number of rows to be displayed.

Say we want to see the last 14 rows, we write the number 14 in the parentheses.

In [54]:
# to display the last 14 rows
df_basic_info.tail(14)

Age  Weight (in kg)  Height (in m)
9    32              21           1.52
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

**8. To obtain the dimension of the data**

In [55]:
# to display the shape of the data
df_basic_info.shape

(23, 3)

**9. To know the data types of a data frame**

In [56]:
# to know the type of each variable
df_basic_info.dtypes

Age                 int64
Weight (in kg)      int64
Height (in m)     float64
dtype: object

We see the data type of each variable.

**10. To know some information of the data**

In [57]:
# to know information on the variables in the data
df_basic_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
Age               23 non-null int64
Weight (in kg)    23 non-null int64
Height (in m)     23 non-null float64
dtypes: float64(1), int64(2)
memory usage: 680.0 bytes


We see this output gives the number of rows present in the data `RangeIndex: 23 entries, 0 to 22` There are 23 rows numbered from 0 to 22. And there are a total of three columns - `Data columns (total 3 columns)`. 

Consider `Age 23 non-null int64` indicates that the column named 'Age' has 23 non-null observations having the data type 'int64'

And finally the memory used to save this dataframe is 680 bytes.

**11. To check the data type of column in the data frame**

In [58]:
# check the type of each variable
type(df_basic_info.Age)

pandas.core.series.Series

In [59]:
# check the type of each variable
type(df_basic_info["Weight (in kg)"])

pandas.core.series.Series

In [60]:
# check the type of each variable
type(df_basic_info["Height (in m)"])

pandas.core.series.Series

**Note that every column of the data frame is a pandas Series.**

<a id="manipulatingDF"> </a>
### 4.2  Manipulating DataFrames 

### Add new column and rows

 **Remark:**   
 1. DataFrame[column] works for any column name, but DataFrame.column only works when the column name is a valid Python variable name.<br>
 2. New columns cannot be created with the ` df_basic_info.BMI ` syntax.
                   

**1. Adding a new column to the data set**

In [61]:
# create a new variable BMI
df_basic_info["BMI"] = df_basic_info["Weight (in kg)"] / df_basic_info["Height (in m)"]**2

In [62]:
# print the DataFrame
df_basic_info.head()

Age  Weight (in kg)  Height (in m)        BMI
0   45              60           1.35  32.921811
1   12              43           1.21  29.369579
2   54              78           1.50  34.666667
3   26              65           1.21  44.395875
4   68              50           1.32  28.696051

In [63]:
# check the shape of the DataFrame
df_basic_info.shape

(23, 4)

**2. Adding a new row to the data set**

A new row can be added using the `loc`

In [64]:
# add a new row at the end of the DataFrame
df_basic_info.loc[24] = [45, 85, 1.8, 26.3]

In [65]:
# display the DataFrame
df_basic_info

Age  Weight (in kg)  Height (in m)        BMI
0   45.0            60.0           1.35  32.921811
1   12.0            43.0           1.21  29.369579
2   54.0            78.0           1.50  34.666667
3   26.0            65.0           1.21  44.395875
4   68.0            50.0           1.32  28.696051
5   21.0            43.0           1.52  18.611496
6   10.0            32.0           1.65  11.753903
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
9   32.0            21.0           1.52   9.089335
10  23.0            53.0           1.50  23.555556
11  34.0            65.0           1.76  20.983988
12  55.0            89.0           1.65  32.690542
13  23.0            45.0           1.75  14.693878
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
16  26.0            65.0           1.21  44.395875
17  56.0            74.0           1.69  25.909457
18  67.0            78.0           1.85  22.790358
19  26.0            65.0           1.21  44.395875
20  68.0            50.0           1.32  28.696051
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358
24  45.0            85.0           1.80  26.300000

We see that a new column number 23 has be added to the data.

**3. Indexing a dataframe using `.iloc`**

`DataFrame.iloc[]` method is used when the index label of a data frame is something other than numeric series of 0, 1, 2, 3….n or in case the user doesn’t know the index label. 

We shall work on the BMI data set.

#### Select the 2nd row

In [66]:
# select the second row
df_basic_info.iloc[2]

Age               54.000000
Weight (in kg)    78.000000
Height (in m)      1.500000
BMI               34.666667
Name: 2, dtype: float64

#### Select fourth, seventh and tenth rows

In [67]:
# select the 4th, 7th and the 10th row
df_basic_info.iloc[[4,7,10]]

Age  Weight (in kg)  Height (in m)        BMI
4   68.0            50.0           1.32  28.696051
7   57.0            34.0           1.61  13.116778
10  23.0            53.0           1.50  23.555556

We use two square brackets since we are passing a list of row numbers to be accessed.

#### Select 12th to 17th rows

In [68]:
# select rows 
df_basic_info.iloc[12:18]

Age  Weight (in kg)  Height (in m)        BMI
12  55.0            89.0           1.65  32.690542
13  23.0            45.0           1.75  14.693878
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
16  26.0            65.0           1.21  44.395875
17  56.0            74.0           1.69  25.909457

#### Select the 1st column

In [69]:
# select the 1st column
df_basic_info.iloc[:, 1]

0     60.0
1     43.0
2     78.0
3     65.0
4     50.0
5     43.0
6     32.0
7     34.0
8     23.0
9     21.0
10    53.0
11    65.0
12    89.0
13    45.0
14    76.0
15    78.0
16    65.0
17    74.0
18    78.0
19    65.0
20    50.0
21    76.0
22    78.0
24    85.0
Name: Weight (in kg), dtype: float64

#### Select the last column

In [70]:
# select the last column
# the colon indicates that all the rows are selected
# -1 indicated that the last column is selected
df_basic_info.iloc[:,-1]

0     32.921811
1     29.369579
2     34.666667
3     44.395875
4     28.696051
5     18.611496
6     11.753903
7     13.116778
8     14.958377
9      9.089335
10    23.555556
11    20.983988
12    32.690542
13    14.693878
14    26.609713
15    22.790358
16    44.395875
17    25.909457
18    22.790358
19    44.395875
20    28.696051
21    26.609713
22    22.790358
24    26.300000
Name: BMI, dtype: float64

To select the last column we use -1, to select the second last column we use -2

#### Select the first two columns

In [71]:
# select the 1st and 2nd columns
# the colon indicates that all the rows are selected
# 0:2 indicates that the 1st and 2nd columns are selected
df_basic_info.iloc[:,0:2]

Age  Weight (in kg)
0   45.0            60.0
1   12.0            43.0
2   54.0            78.0
3   26.0            65.0
4   68.0            50.0
5   21.0            43.0
6   10.0            32.0
7   57.0            34.0
8   75.0            23.0
9   32.0            21.0
10  23.0            53.0
11  34.0            65.0
12  55.0            89.0
13  23.0            45.0
14  56.0            76.0
15  67.0            78.0
16  26.0            65.0
17  56.0            74.0
18  67.0            78.0
19  26.0            65.0
20  68.0            50.0
21  56.0            76.0
22  67.0            78.0
24  45.0            85.0

#### Select the first two columns and 5 to 10 rows

In [72]:
# 5:11 indicates that the 5th to 10th rows will be selected
# 0:2 indicates that the 1st and 2nd columns be selected
df_basic_info.iloc[5:11, 0:2]

Age  Weight (in kg)
5   21.0            43.0
6   10.0            32.0
7   57.0            34.0
8   75.0            23.0
9   32.0            21.0
10  23.0            53.0

**4. Indexing a dataframe using `.loc`**

`DataFrame.loc[]` method is a method that takes only index labels and returns row or dataframe if the index label exists in the caller data frame. <br>
`DataFrame.loc[Row_names, column_names]` is used to select or index rows or columns based on their name.

#### Select 1 to 5 rows and 2nd and 4th columns

In [73]:
# 1:5 indicates that rows with indices 1, 2, 3, 4 and 5 are selected
# ["Weight (in kg)","BMI"] indicates that the specified columns be selected
df_basic_info.loc[1:5,["Weight (in kg)","BMI"]]

Weight (in kg)        BMI
1            43.0  29.369579
2            78.0  34.666667
3            65.0  44.395875
4            50.0  28.696051
5            43.0  18.611496

**Note:** the row names are numbers 

**5. Selecting columns by specifying column names**

#### Select the column 'Age'

In [74]:
# select the coumn 'Age'
df_basic_info.Age

0     45.0
1     12.0
2     54.0
3     26.0
4     68.0
5     21.0
6     10.0
7     57.0
8     75.0
9     32.0
10    23.0
11    34.0
12    55.0
13    23.0
14    56.0
15    67.0
16    26.0
17    56.0
18    67.0
19    26.0
20    68.0
21    56.0
22    67.0
24    45.0
Name: Age, dtype: float64

**Remark:** Using this method we can select only one column.

In [75]:
# OR
df_basic_info["Age"]

0     45.0
1     12.0
2     54.0
3     26.0
4     68.0
5     21.0
6     10.0
7     57.0
8     75.0
9     32.0
10    23.0
11    34.0
12    55.0
13    23.0
14    56.0
15    67.0
16    26.0
17    56.0
18    67.0
19    26.0
20    68.0
21    56.0
22    67.0
24    45.0
Name: Age, dtype: float64

#### Select the column 'Age' and 'BMI'

In [76]:
# select two columns
# the column names are passed in a list
df_basic_info[["Age","BMI"]]

Age        BMI
0   45.0  32.921811
1   12.0  29.369579
2   54.0  34.666667
3   26.0  44.395875
4   68.0  28.696051
5   21.0  18.611496
6   10.0  11.753903
7   57.0  13.116778
8   75.0  14.958377
9   32.0   9.089335
10  23.0  23.555556
11  34.0  20.983988
12  55.0  32.690542
13  23.0  14.693878
14  56.0  26.609713
15  67.0  22.790358
16  26.0  44.395875
17  56.0  25.909457
18  67.0  22.790358
19  26.0  44.395875
20  68.0  28.696051
21  56.0  26.609713
22  67.0  22.790358
24  45.0  26.300000

**6. Conditional subsetting**

Selecting rows where the value of `Age` is greater than 47 

In [77]:
# to selct rows where the Age is greater than 47
df_basic_info[df_basic_info['Age'] > 47]

Age  Weight (in kg)  Height (in m)        BMI
2   54.0            78.0           1.50  34.666667
4   68.0            50.0           1.32  28.696051
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
12  55.0            89.0           1.65  32.690542
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
17  56.0            74.0           1.69  25.909457
18  67.0            78.0           1.85  22.790358
20  68.0            50.0           1.32  28.696051
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358

Subsetting the age more than 40 *or* weight column value more than 65

In [78]:
# to select rows where either age is greater than 40 or weight is more than 65
df_basic_info[ (df_basic_info["Age"] > 40) | (df_basic_info['Weight (in kg)'] > 65)]

Age  Weight (in kg)  Height (in m)        BMI
0   45.0            60.0           1.35  32.921811
2   54.0            78.0           1.50  34.666667
4   68.0            50.0           1.32  28.696051
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
12  55.0            89.0           1.65  32.690542
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
17  56.0            74.0           1.69  25.909457
18  67.0            78.0           1.85  22.790358
20  68.0            50.0           1.32  28.696051
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358
24  45.0            85.0           1.80  26.300000

Subsetting the age *and* weight column value more than 65

In [79]:
# select rows where both age and weight are more than 65
df_basic_info[(df_basic_info["Age"] > 65) & (df_basic_info['Weight (in kg)'] > 65)]

Age  Weight (in kg)  Height (in m)        BMI
15  67.0            78.0           1.85  22.790358
18  67.0            78.0           1.85  22.790358
22  67.0            78.0           1.85  22.790358

**7. Sort the data frame on the basis of values in a column**

Each column of a pandas DataFrame is treated as a pandas Series. The `.sort_values()` in DataFrames works similar to the `pandas.Series`.


In [80]:
# sort the data frame on basis of 'Age' values
# by default the values will get sorted in ascending order
df_basic_info.sort_values('Age')

# Note: 'ascending = False' will sort the data frame in descending order

Age  Weight (in kg)  Height (in m)        BMI
6   10.0            32.0           1.65  11.753903
1   12.0            43.0           1.21  29.369579
5   21.0            43.0           1.52  18.611496
13  23.0            45.0           1.75  14.693878
10  23.0            53.0           1.50  23.555556
19  26.0            65.0           1.21  44.395875
3   26.0            65.0           1.21  44.395875
16  26.0            65.0           1.21  44.395875
9   32.0            21.0           1.52   9.089335
11  34.0            65.0           1.76  20.983988
0   45.0            60.0           1.35  32.921811
24  45.0            85.0           1.80  26.300000
2   54.0            78.0           1.50  34.666667
12  55.0            89.0           1.65  32.690542
14  56.0            76.0           1.69  26.609713
17  56.0            74.0           1.69  25.909457
21  56.0            76.0           1.69  26.609713
7   57.0            34.0           1.61  13.116778
22  67.0            78.0           1.85  22.790358
15  67.0            78.0           1.85  22.790358
18  67.0            78.0           1.85  22.790358
4   68.0            50.0           1.32  28.696051
20  68.0            50.0           1.32  28.696051
8   75.0            23.0           1.24  14.958377

**8. Rank the dataframe**

In [81]:
# rank the data frame 'data' in descending order based on 'BMI'
# 'method = min' assigns the minimum rank to highest equal value of 'BMI' 
df_basic_info['Rank_min'] = df_basic_info['BMI'].rank(ascending = 0, method  = 'min')

# print the data
df_basic_info

Age  Weight (in kg)  Height (in m)        BMI  Rank_min
0   45.0            60.0           1.35  32.921811       5.0
1   12.0            43.0           1.21  29.369579       7.0
2   54.0            78.0           1.50  34.666667       4.0
3   26.0            65.0           1.21  44.395875       1.0
4   68.0            50.0           1.32  28.696051       8.0
5   21.0            43.0           1.52  18.611496      19.0
6   10.0            32.0           1.65  11.753903      23.0
7   57.0            34.0           1.61  13.116778      22.0
8   75.0            23.0           1.24  14.958377      20.0
9   32.0            21.0           1.52   9.089335      24.0
10  23.0            53.0           1.50  23.555556      14.0
11  34.0            65.0           1.76  20.983988      18.0
12  55.0            89.0           1.65  32.690542       6.0
13  23.0            45.0           1.75  14.693878      21.0
14  56.0            76.0           1.69  26.609713      10.0
15  67.0            78.0           1.85  22.790358      15.0
16  26.0            65.0           1.21  44.395875       1.0
17  56.0            74.0           1.69  25.909457      13.0
18  67.0            78.0           1.85  22.790358      15.0
19  26.0            65.0           1.21  44.395875       1.0
20  68.0            50.0           1.32  28.696051       8.0
21  56.0            76.0           1.69  26.609713      10.0
22  67.0            78.0           1.85  22.790358      15.0
24  45.0            85.0           1.80  26.300000      12.0

From the above data frame, we can see that 'BMI = 44.395875' is repeating thrice; thus the method = 'min' will assign the minimum rank (=1) to all the three values of BMI. The rank '4' will be assigned to the second largest value of BMI and so on. Thus, there is no rank equal to 2 and 3.

In [82]:
# method = 'dense' assigns same rank to all the same BMI values
df_basic_info['Rank_densed'] = df_basic_info['BMI'].rank(method = 'dense')

# print the data
df_basic_info

Age  Weight (in kg)  Height (in m)        BMI  Rank_min  Rank_densed
0   45.0            60.0           1.35  32.921811       5.0         16.0
1   12.0            43.0           1.21  29.369579       7.0         14.0
2   54.0            78.0           1.50  34.666667       4.0         17.0
3   26.0            65.0           1.21  44.395875       1.0         18.0
4   68.0            50.0           1.32  28.696051       8.0         13.0
5   21.0            43.0           1.52  18.611496      19.0          6.0
6   10.0            32.0           1.65  11.753903      23.0          2.0
7   57.0            34.0           1.61  13.116778      22.0          3.0
8   75.0            23.0           1.24  14.958377      20.0          5.0
9   32.0            21.0           1.52   9.089335      24.0          1.0
10  23.0            53.0           1.50  23.555556      14.0          9.0
11  34.0            65.0           1.76  20.983988      18.0          7.0
12  55.0            89.0           1.65  32.690542       6.0         15.0
13  23.0            45.0           1.75  14.693878      21.0          4.0
14  56.0            76.0           1.69  26.609713      10.0         12.0
15  67.0            78.0           1.85  22.790358      15.0          8.0
16  26.0            65.0           1.21  44.395875       1.0         18.0
17  56.0            74.0           1.69  25.909457      13.0         10.0
18  67.0            78.0           1.85  22.790358      15.0          8.0
19  26.0            65.0           1.21  44.395875       1.0         18.0
20  68.0            50.0           1.32  28.696051       8.0         13.0
21  56.0            76.0           1.69  26.609713      10.0         12.0
22  67.0            78.0           1.85  22.790358      15.0          8.0
24  45.0            85.0           1.80  26.300000      12.0         11.0

Here, the dense method assigns minimum rank (=1) to the minimum value (=9.089335) of the BMI. Rank 2 will be assigned to next smallest value and so on. The value 44.395875 which repeats thrice has the same rank - 18 to the three observations.

**9. To check for missing values**

We shall import a new dataset.

In [83]:
# read the dataset
df_basic_info_missing = pd.read_csv("basic_info_missingdata.csv")

# print the dataset
df_basic_info_missing

Age  Weight (in kg)  Height (in m)
0   45.0            60.0           1.35
1   12.0            43.0           1.21
2   54.0            78.0           1.50
3   26.0            65.0           1.21
4   68.0            50.0           1.32
5   21.0             NaN           1.52
6   10.0            32.0           1.65
7   57.0            34.0           1.61
8   75.0            23.0           1.24
9   32.0            21.0           1.52
10  23.0             NaN           1.50
11  34.0            65.0           1.76
12  55.0            89.0           1.65
13  23.0            45.0           1.75
14  56.0            76.0           1.69
15  67.0            78.0           1.85
16  26.0            65.0           1.21
17  56.0            74.0           1.69
18  67.0            78.0            NaN
19  26.0            65.0           1.21
20  68.0            50.0           1.32
21   NaN            76.0           1.69
22  67.0            78.0           1.85

Check for missing values

In [84]:
# isnull() returns True for a missing value
# sum() gives the sum of True values
df_basic_info_missing.isnull().sum()

Age               1
Weight (in kg)    2
Height (in m)     1
dtype: int64

The function `.isnull` check whether the data is missing. The `sum()` sums the number of 'True' values in the column. The final output gives the number of missing values in each column.

Here, we see there are 2 missing values in the 'weight' column and one missing value in other columns.

<a id="reading_data"> </a>
## 5.  Reading Data from Different Sources

**1. Read a `.xlsx` file**

In [85]:
# to read a xlsx file
pd.read_excel('employee_info.xlsx')

Age  Gender  Salary
0   45    Male   40000
1   12    Male       0
2   54  Female  150000
3   26    Male   30000
4   64  Female   15000
5   21  Female   25600

**2. Read a `.zip` file**

The zipped file contain a .csv file

In [86]:
# import the library zipfile
import zipfile

# Zipfile reads the zipped file
# from the zipped file open the csv as f
# read the csv file
with zipfile.ZipFile('data.zip') as z:
    with z.open('example.csv') as f:
        file = pd.read_csv(f)
        print(file.head())

   Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50
3   26              65           1.21
4   68              50           1.32


**3. Read a `.html` file**

In [87]:
# read the html using read_html()
# header = 1 indicates that the first row contains the headings
# index_col = 0 indicates that the first column contains the index
df_emp = pd.read_html('employee.html', header=1, index_col=0)

# print the data
df_emp

[   Age  Gender  Salary
 1                     
 2   45    Male   40000
 3   12    Male       0
 4   54  Female  150000
 5   26    Male   30000
 6   64  Female   15000
 7   21  Female   25600]

**4. Read a `.txt` file**

In [88]:
# read the text file
# sep is the delimiter use
df_demography = pd.read_csv('demography.txt', sep="\t")

# print the head of the data
df_demography.head()

Country  Birth rate  Life expectancy
0      Vietnam       1.822        74.828244
1      Vanuatu       3.869        70.819488
2        Tonga       3.911        72.150659
3  Timor-Leste       5.578        61.999854
4     Thailand       1.579        73.927659

**5. Read a `.json` file**

In [89]:
# read the file
pd.read_json('iris.json')

sepalLength  sepalWidth  petalLength  petalWidth    species
0            5.1         3.5          1.4         0.2     setosa
1            4.9         3.0          1.4         0.2     setosa
2            4.7         3.2          1.3         0.2     setosa
3            4.6         3.1          1.5         0.2     setosa
4            5.0         3.6          1.4         0.2     setosa
..           ...         ...          ...         ...        ...
145          6.7         3.0          5.2         2.3  virginica
146          6.3         2.5          5.0         1.9  virginica
147          6.5         3.0          5.2         2.0  virginica
148          6.2         3.4          5.4         2.3  virginica
149          5.9         3.0          5.1         1.8  virginica

[150 rows x 5 columns]

**6. Read a `.xml` file**

In [90]:
# import the library to xml file
import xml.etree.ElementTree as ET 

# extract the xml file
tree = ET.parse("student_data.xml")
root = tree.getroot() 

# assign column names of the output DataFrame
df_col = ["Name", "Gender", "Marks"]

# create an empty list 
rows = []

for node in root: 
    name = node.attrib.get("name")
    gender = node.find("gender").text if node is not None else None
    marks = node.find("marks").text if node is not None else None
    
    # append each observation in the data to ‘rows’
    rows.append({"Name": name, "Gender": gender, 
                 "Marks": marks})

# create a DataFrame    
df_student = pd.DataFrame(rows, columns = df_col)

# print the DataFrame
df_student

Name  Gender Marks
0  Student_1  Female   9.5
1  Student_2    Male   8.2
2  Student_3    Male   7.8
3  Student_4  Female   8.5